bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+*** Let ''t<sub>n</sub> = sha256(ecdh_shared_secret || ser<sub>32</sub>(n))''
+*** Let ''P<sub>mn</sub> = B<sub>m</sub> + t<sub>n</sub>·G''
+*** Encode ''P<sub>mn</sub>'' as a [https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki BIP341] taproot output
+*** Optionally, repeat with n++ to create additional outputs for the current ''B<sub>m</sub>''
+*** If no additional outputs are required, continue to the next ''B<sub>m</sub>'' with ''n++''<ref name="why_not_the_same_tn">''' Why not re-use ''t<sub>n</sub>''? when paying different labels to the same receiver?''' If paying the same entity but to two separate labeled addresses in the same transaction without incrementing ''n'', the two outputs would be ''B<sub>spend</sub> + t<sub>n</sub>·G + i·G'' and ''B<sub>spend</sub> + t<sub>n</sub>·G + j·G''. The attacker could subtract the two values and observe that the distance between i and j is small. This would allow them to deduce that this transaction is a silent payment transaction and that a single entity received two outputs, but won't tell them who the entity is.</ref>
+
+=== Receiver ===
+
+==== Key Derivation ====
+
+Two keys are needed to create a silent payments address: the spend key and the scan key. While these keys can be generated independently, wallet software SHOULD use BIP32 derivation<ref name="bip32_derivation">'''Why use BIP32 hardened derivation?''' Using BIP32 derivation allows users to add silent payments to an existing master seed. It also ensures that a user's silent payment funds are recoverable in any wallet which supports BIP32 derivation. Using hardened derivation ensures that it is safe to export the scan private key without exposing the master key or spend private key.</ref> to ensure compatibility across wallets.
+
+A scan and spend key pair using BIP32 derivation are defined (taking inspiration from [https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki BIP44]) in the following manner:
+
+
+    Spend private key: m / purpose' / coin_type' / account' / silent_payment_version' / 0' / 0
+     Scan private key: m / purpose' / coin_type' / account' / silent_payment_version' / 1' / 0
+
+Wallet software MUST use hardened derivation to ensure the master key is not exposed in the event the scan private key is compromised. Purpose is a constant set to ''BIP_NUMBER'' following the BIP43 recommendation. Refer to [https://github.com/bitcoin/bips/blob/master/bip-0043.mediawiki BIP43] and [https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki BIP44] for more details.
+
+==== Scanning ====
+
+For each transaction the receiving wallet suspects might be a silent payment to themselves, it must:
+
+* Generate the ''outpoints_hash'', using the method described above
+* Let ''A = A<sub>0</sub> + A<sub>1</sub> + … A<sub>n</sub>'', where each ''A<sub>i</sub>'' is the public key of an input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
```

Comment:
In my own implementation I'm running into parity issues: because _B<sub>scan</sub>_ is a BIP340 public key which always has an even parity while _A_ is a sum of potentially non-BIP340 keys and can thus have any parity, in about half the cases _ecdh_shared_secret_ is different for sender and receiver. Isn't there a step missing that corrects for this, or am I missing something?

Reply:
Interesting, do you have some test data I can use to try and reproduce this? I _think_ mixing of taproot and non-taproot is covered in the functional tests, so perhaps there is a step we are doing in the Bitcoin Core implementation that isn't explained clearly in the BIP.

Reply:
Here are the test vectors that cause a parity mismatch in my implementation, along with the computed shared secret and partial computations.

_B<sub>scan</sub>_ = 76e246e6b92aec2efd6c8db8037d284651ccd3c9b9ac1404010b7b3268016e92
_b<sub>scan</sub>_ = ad8af2bd1bf108f3dd2ac74a2a5ec51d9a173c684c9e6365c6af93d5c5fdf61a
_B<sub>spend</sub>_ = _B<sub>m</sub>_ = 8810c8029bb9560ab5c54241a108912f847ab7975df626f816430a64a35be68c
_A<sub>0</sub>_ = _A_ = 03603ce64dcf386cd11c97b6c6d94de719c4638101ae9613f2f1e6fae78bb89e0f
_a<sub>0</sub>_ = _a_ = 5f230aed01b015c4fdd78d209f878cb67be5135504b4d7a2c85b368d1c042b9d
_outpoints_hash_ = 2f11ade1789049b1a26837d517bae661fb3b2dfc00a0d812f65ee207c7617c05

**Sender:**
_a_·_B<sub>scan</sub>_ = 03bc2412cc4aad38e93f0ce1ad5628ae231aac7cf1b6f54dfb074999bd34076f68
_ecdh_shared_secret_ = 028408231298cc86c22e61ca8ba1967dd9192d68bcb4ca575e7f8284ae9c417828

**Receiver:**
_b<sub>scan</sub>_·_A_ = 02bc2412cc4aad38e93f0ce1ad5628ae231aac7cf1b6f54dfb074999bd34076f68
_ecdh_shared_secret_ = 038408231298cc86c22e61ca8ba1967dd9192d68bcb4ca575e7f8284ae9c417828

Reply:
Thanks for the test data! Your parity mismatch is coming from the fact that _b<sub>scan</sub>_ produces an odd y value and needs to be negated.

It was my understanding that in order to be a valid BIP340 public key, the private key must be negated if it returns an odd y. Rereading BIP340, however, this doesn't seem to be clearly specified. The closest reference I can find is:

> Note that we use a very different public key format (32 bytes) than the ones used by existing systems (which typically use elliptic curve points as public keys, or 33-byte or 65-byte encodings of them). A side effect is that PubKey(sk) = PubKey(bytes(n - int(sk)), so every public key has two corresponding secret keys. 

So I think we do need to specify this in the silent payments BIP. Perhaps something like: 

> "if `has_even_y(P)` is false, negate _b<sub>scan</sub>_"

Reply:
I don't have a strong opinion on how to correct the parity mismatch, but "if _has_even_y(B<sub>scan</sub>)_ is false, negate _b<sub>scan</sub>_" has the same problem as mentioned below that B<sub>scan</sub> is not an EC point but a byte array. [BIP340: Default Signing](https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#default-signing) includes notation for private key negation which you could reuse. Another option would be to use 33-byte keys instead of x-only keys (I suppose this choice was made to make the silent payment address shorter?).

Reply:
I wrote some test cases for this and turns out it's very difficult to keep track of the parity between sender and receiver, especially when using labels. While it could be done with x-only public keys, it would involve checking the evenness and negating keys at almost every step. I think the extra byte is definitely worth adding to keep things simple and save on computation when scanning.

I've updated the specification to use 33-byte keys for the address, which I believe also addresses the notation concern.

Reply:
Additionally, your example made me notice that we need to check both _output - P<sub>n</sub>_ and _output\_negated - P<sub>n</sub>_ when scanning for labels. This wasn't necessary when using the old method of brute-forcing the label checks for all _m_, which is why we didn't notice it when we updated to use the subtraction technique when scanning for labels. I also updated the BIP to include this step. Thanks so much for working on an implementation and helping uncover this with your test case @vostrnad !

Reply:
I've looked over the changes and they seem to mostly address the parity issues, however it seems to me that one step is missing: doesn't the sender need to negate the taproot private keys whose public keys have an odd Y coordinate?

Reply:
Yep, that's correct. I had always assumed that being a taproot outpoint meant you always used the private key which corresponds with the even Y coordinate, but seems we should explicitly spell that out as a step in the BIP.

Reply:
Added a sentence to mention checking and negating taproot keys, if necessary